pwem/protocols/protocol_sets.py

```diff
@@ -884,3 +884,84 @@ def _summary(self):
                        ' particles.' % (self.outputParticles.getSize(),
                                         self.inputParticles.get().getSize())]
         return summary
+
+
+class ProtSubSetRange(ProtSets):
+    """
+    Create a subset from a set based on a range.
+
+    This overlaps with more general protocol subset 'range' option, but this one
+    is more specific and much faster.
+    """
+    _label = 'subset range'
+
+    # -------------------------- DEFINE param functions -----------------------
+    def _defineParams(self, form):
+        form.addSection(label='Input')
+
+        add = form.addParam  # short notation
+        add('inputSet', pwprot.params.PointerParam, pointerClass='EMSet',
+            label="Input set of items", important=True,
+            help='Input from where the subset will be created.')
+        add('rangeFrom', pwprot.params.IntParam, default=1,
+            label="From",
+            help='Starting index (1-based)')
+        add('rangeTo', pwprot.params.IntParam, default=-1,
+            label="To",
+            help='Ending index (-1 means last index).')
+
+    # -------------------------- INSERT steps functions -----------------------
+    def _insertAllSteps(self):
+        self._insertFunctionStep('createOutputStep')
+
+    # -------------------------- STEPS functions ------------------------------
+    def createOutputStep(self):
+        inputSet = self.inputSet.get()
+
+        inputClassName = inputSet.getClassName()
+
+        try:
+            outputSetFunction = getattr(self, "_create%s" % inputClassName)
+            outputSet = outputSetFunction()
+        except Exception:
+            outputSet = inputSet.createCopy(self._getPath())
+
+        outputSet.copyInfo(inputSet)
+
+        start = self.rangeFrom.get() - 1
+        rangeTo = self.rangeTo.get()
+        end = (rangeTo if rangeTo >= 0 else inputSet.getSize()) - 1
+
+        for i, item in enumerate(inputSet.iterItems()):
+            if i > end:
+                break
+            if i >= start:
+                outputSet.append(item)
+
+        if n := outputSet.getSize():
+            key = 'output' + inputClassName.replace('SetOf', '')
```

Comment:
With this type of protocols, (subsets, joins,...) It is almost impossible to define possible outputs with this naming scheme. My suggestion would be to choose a fixed name and define possible outputs so it can be used when planning a workflow.

Reply:
Hi Pablo, as you said, it is almost impossible...since it can create subsets of many set types. So, I prefer to not define any possible output because it will add many "possible" entries to the selection dialog. I think this protocol is mostly used through API, where the workflow can be built anyway...or more in normal GUI processing one step after the other.